taichi/backends/cuda/cuda_profiler.cpp

```diff
@@ -0,0 +1,150 @@
+    profiler_config_.enable = true;
+    profiler_config_.profiler_type = CUDA_KERNEL_PROFILER_EVENT;
+    profiler_config_.profiling_mode = KernelProfilerMode::enable;
+    TI_TRACE("CUDA_KERNEL_PROFILER_EVENT : enable");
+    TI_TRACE("CUDAProfiler enable = {}", profiler_config_.enable);
+    TI_TRACE("profiler_type : {}", profiler_config_.profiler_type);
+    return true;
+  }
+
+#if !defined(TI_WITH_TOOLKIT_CUDA)
+  if (enable && profiler_type == CUDA_KERNEL_PROFILER_CUPTI) {
+    TI_WARN(
+        "CUPTI toolkit is not compiled with taichi, fallback to cuEvent kernel "
+        "profiler");
+    TI_WARN(
+        "to use CUPTI kernel profiler : "
+        "TAICHI_CMAKE_ARGS=-DTI_WITH_TOOLKIT_CUDA=True python3 setup.py "
+        "develop --user");
+
+    profiler_config_.enable = true;
+    profiler_config_.profiler_type = CUDA_KERNEL_PROFILER_EVENT;
+    profiler_config_.profiling_mode = KernelProfilerMode::enable;
+
+    TI_TRACE("CUDA_KERNEL_PROFILER_EVENT : enable");
+    TI_TRACE("CUDAProfiler enable = {}", profiler_config_.enable);
+    TI_TRACE("profiler_type : {}", profiler_config_.profiler_type);
+    return true;
+  }
+#else
+// TODO::CUPTI_PROFILER
+#endif
+}
+
+bool CUDAProfiler::statisticsOnRecords(
+    std::vector<KernelProfileRecord> &records,
+    double &total_time_ms) {
+  size_t result_nums = traced_records_.size();
+  if (!result_nums) {
+    TI_WARN("traced_records_ is empty!");
+    return false;
+  }
+
+  for (size_t resultIndex = 0; resultIndex < result_nums; ++resultIndex) {
+    auto it = std::find_if(
+        records.begin(), records.end(), [&](KernelProfileRecord &r) {
+          return r.name == traced_records_[resultIndex].kernel_name;
+        });
+    if (it == records.end()) {
+      records.push_back(traced_records_[resultIndex].kernel_name);
+      it = std::prev(records.end());
+    }
+    it->insert_sample(traced_records_[resultIndex].kernel_elapsed_time_in_ms);
+    total_time_ms += traced_records_[resultIndex].kernel_elapsed_time_in_ms;
+
+    if (profiler_config_.profiling_mode == KernelProfilerMode::cuda_accurate) {
+      it->cuda_global_access(
+          traced_records_[resultIndex].kernel_gloabl_load_byets,
+          traced_records_[resultIndex].kernel_gloabl_store_byets);
+    }
+    if (profiler_config_.profiling_mode == KernelProfilerMode::cuda_detailed) {
+      it->cuda_global_access(
+          traced_records_[resultIndex].kernel_gloabl_load_byets,
+          traced_records_[resultIndex].kernel_gloabl_store_byets);
+      it->cuda_uti_ratio(traced_records_[resultIndex].utilization_ratio_sm,
+                         traced_records_[resultIndex].utilization_ratio_mem);
+    }
+  }
+
+  return true;
+}
+
+CUDAProfiler &CUDAProfiler::get_instance_without_context() {
+  static CUDAProfiler *instance = new CUDAProfiler();
+  return *instance;
+}
+
+CUDAProfiler &CUDAProfiler::get_instance() {
+  CUDAContext::get_instance();
+  return get_instance_without_context();
```

Comment:
Any specific reason we need to do this? If possible, I'd like us to avoid this

Reply:
Solved in commit f4db404, thanks !